src/storage/src/hummock/shared_buffer/shared_buffer_manager.rs

```diff
@@ -14,253 +14,137 @@
         Self {
-            shared_buffer: PLRwLock::new(BTreeMap::new()),
+            capacity: options.shared_buffer_capacity as usize,
+            core,
             uploader_tx,
             uploader_handle,
-            stats,
-            ongoing_flush: OnGoingFlush::new(),
+            flush_notifier,
         }
     }
 
-    fn empty(&self) -> bool {
-        self.stats.shared_buffer_cur_size.load(Ordering::SeqCst) == 0
+    pub fn size(&self) -> usize {
+        self.core.read().size
     }
 
-    /// Allocates a shared buffer budget.
-    async fn allocate_space(&self, batch_size: u64) -> HummockResult<()> {
-        let mut current_size = self.stats.shared_buffer_cur_size.load(Ordering::SeqCst);
-        let threshold = self.stats.shared_buffer_threshold_size;
-
-        // Atomically allocates space,
-        // since there could be concurrent Actors write to the shared buffer.
-        'retry_allocate: loop {
-            // flush shared buffer if there is no enough space
-            while threshold < current_size + batch_size {
-                log::debug!(
-                    "trigge flush: threshold {}, new_size {}",
-                    threshold,
-                    current_size + batch_size
-                );
-                self.flush().await?;
-                current_size = self.stats.shared_buffer_cur_size.load(Ordering::SeqCst);
-            }
-
-            let res = self.stats.shared_buffer_cur_size.compare_exchange(
-                current_size,
-                current_size + batch_size,
-                Ordering::SeqCst,
-                Ordering::Acquire,
-            );
-            match res {
-                Ok(_) => {
-                    assert!(current_size + batch_size <= threshold);
-                    break; // success
-                }
-                Err(old_val) => {
-                    current_size = old_val;
-                    continue 'retry_allocate;
-                }
-            }
-        }
-        Ok(())
+    pub fn is_empty(&self) -> bool {
+        self.size() == 0
     }
 
     /// Puts a write batch into shared buffer. The batch will be synced to S3 asynchronously.
     pub async fn write_batch(
         &self,
-        batch: Vec<SharedBufferItem>,
+        items: Vec<SharedBufferItem>,
         epoch: HummockEpoch,
     ) -> HummockResult<u64> {
-        let batch = SharedBufferBatch::new(batch, epoch);
-        let size = batch.size;
-
-        self.allocate_space(size).await?;
+        let batch = SharedBufferBatch::new(items, epoch);
+        let batch_size = batch.size;
+
+        // Stall writes that attempting to exceeds capacity.
+        // Actively trigger flush, suspend, and wait until some flush completed.
+        while self.size() + batch_size as usize > self.capacity {
+            self.uploader_tx
+                .send(UploadItem::Flush)
+                .map_err(HummockError::shared_buffer_error)?;
+            self.flush_notifier.notified().await;
```

Comment:
I found a potential bug in multiple concurrent `write_batch` scenario. 

Let's assume that there are two concurrent writers, both enter this while loop, and have sent the `UploadItem::Flush` to the uploader channel, but neither of them has called `notified`. And then the uploader might have finished processing the two `UploadItem::Flush`s and called the `flush_notifier.notify_one` for at least twice. 

However, as the doc of [`tokio::sync::Notify::notify_one`](https://docs.rs/tokio/1.0.2/tokio/sync/struct.Notify.html#method.notify_one) mentioned,
> At most one permit may be stored by Notify. Many sequential calls to notify_one will result in a single permit being stored.

Therefore, only one permit in the notify will be stored. And then when the two writers run `self.flush_notifier.notified()`, only one can make progress, and the other may be hung forever.

Reply:
Agreed.